syntax = "proto3";

service OrderBookService {
  rpc AddOrder    ( OrderRequest ) returns ( OrderResponse );
  rpc GetBestBid  ( Empty        ) returns ( OrderResponse );
  rpc GetBestAsk  ( Empty        ) returns ( OrderResponse );
  rpc GetOrderBook( Empty        ) returns ( OrderBookResponse );
  rpc GetTradeLog ( Empty        ) returns ( stream Trade );
}

message OrderRequest {
  int32  order_id   = 1;
  double price      = 2;
  uint32 quantity   = 3;
  bool   side       = 4;  // true=buy, false=sell
  int64  timestamp  = 5;
  string order_type = 6;  // "LIMIT" or "MARKET"
}

message OrderResponse {
  int32  order_id   = 1;
  double price      = 2;
  uint32 quantity   = 3;
  bool   side       = 4;
  int64  timestamp  = 5;
  string order_type = 6;
}

message Empty {}

message OrderBookLevel {
    double price = 1;
    uint32 quantity = 2;
}

message OrderBookResponse {
    repeated OrderBookLevel bids = 1;
    repeated OrderBookLevel asks = 2;
}

message Trade {
    int64 timestamp = 1;
    double price = 2;
    uint32 quantity = 3;
    bool side = 4;
}
